concise/callbacks.py

```diff
@@ -0,0 +1,108 @@
+""" EvalMetric Callback: utility called at certain points during model training.
+Author: Mohammadamin Barekatain
+Affiliation: TUM
+"""
+
+import keras
+import numpy as np
+
+
+class EvalMetric(keras.callbacks.Callback):
+
+    """Evaluate the single-task or multi-task model after every epoch via the given metric.
+    Evaluation is done for each individual task and the average is also reported.
+    Evaluation results are added to logs so that they can be used with
+    other callbacks such as TensorBoard, EarlyStopping, ModelCheckpoint
+    when this callback is added to callback list earlier.
+    # Arguments
+        tasks: list of strings, task names which is used to retrieve the labels for each task.
+        eval_metric: function(y_true, y_pred) which is used for evaluation.
+        eval_metric_name: string, name of the evaluation metric. This is used when reporting the evaluations.
+        num_input_fea: int, number of inputs that the model gets.
+        verbose: verbosity mode, 0 or 1.
+        train_data: If None, the evaluation is done only on validation data.
+    """
+
+    def __init__(self, tasks, eval_metric, eval_metric_name, num_input_fea, verbose=1, train_data=None):
```

Comment:
I would change the arguments in the following way:
- `metrics` - single function/string, a list or a dictionary of functions/strings. In case a list or a simple function are provided, infer the metric names from the function name. Basically, mimic the API of `concise.hyopt.CompileFN`
  - https://github.com/gagneurlab/concise/blob/master/concise/hyopt.py#L418-L423
  - https://github.com/gagneurlab/concise/blob/master/concise/hyopt.py#L471-L475
  - https://github.com/gagneurlab/concise/blob/master/concise/hyopt.py#L729
- `task_names`=None - optional list of task_names. If None, incremental integer is used
- `verbose`:  verbosity mode, 0 or 1. 1 gives a more verbose output.
- `train_data` - Training data used for evaluation.  If None, the evaluation is done only on validation data.

## Other notes
- `num_input_fea` should be automatically inferred from line 62